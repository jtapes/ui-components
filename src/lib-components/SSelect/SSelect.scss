@import '../../theme/styles/variables';
@import '../../theme/styles/mixins';
@import '../../theme/styles/reset';
@import '../../theme/styles/fonts';
@import '../../theme/styles/global';

.s-select {
  position: relative;
  padding-top: 3px;
  width: 100%;
  height: $size--48;

  &__wrapper {
    position: relative;
    display: flex;
    flex-direction: column;
    border-bottom: 1px solid $color--border;
    width: 100%;
    min-height: $size--48;
    background-color: $color--white;
    transition: $animate--time;
  }

  &__dropdown {
    @include z-index(modal);

    position: absolute;
    left: -1px;
    right: -1px;
    top: 100%;
    border-top: 1px solid $color--border;
    border-radius: 0 0 $size--24 $size--24;
    padding-bottom: $size--20;
    box-shadow: 0 3px 0 $color--outline, 3px 3px 0 $color--outline, -3px 3px 0 $color--outline;
    background-color: #fff;

    &--appended-to-body {
      right: auto;
    }

    &--opened {
      padding-top: 0;

      .s-select {
        &__items {
          overflow: auto;

          &::-webkit-scrollbar {
            width: $size--8;
            background-color: transparent;
          }

          &::-webkit-scrollbar-thumb {
            border-radius: $size--4;
            background-color: $color--font-muted;

            &:hover {
              background-color: $color--font-secondary;
            }
          }
        }
      }
    }
  }

  &__toggler {
    cursor: pointer;

    .s-input {
      border-color: transparent;
      background-color: transparent;

      &__append {
        margin-right: -$size--8;
      }

      &:focus-within {
        box-shadow: none;
      }

      &:hover {
        border-color: transparent;
      }

      &.s-input--rounded {
        .s-input__inner {
          border-radius: 0;
          padding-right: 0;
          padding-left: 0;
        }
      }

      &.s-input--readonly {
        .s-input__placeholder:not(.s-input__placeholder--focused) {
          color: $color--font;
        }

        .s-input__inner {
          color: $color--font;
          background-color: transparent;
        }
      }
    }

    .svg-icon {
      transition: $animate--time;
    }
  }

  &__search {
    padding: 0 $size--16;

    .s-input {
      border: 0;
      border-bottom: 1px solid $color--border;
      border-radius: 0;

      &__append {
        margin-right: -$size--8;
      }

      &:focus-within {
        box-shadow: none;
      }

      &.s-input--rounded {
        .s-input {
          &__inner {
            padding-right: 0;
            padding-left: 0;
          }
        }
      }
    }
  }

  &__placeholder {
    position: absolute;
    transition: $animate--time;

    &--focused {
      @include text-xs;

      color: $color--font-muted;
      transform: translateY(-100%);
    }
  }

  &__empty {
    padding: $size--12 $size--16;
    max-height: $size--40;
  }

  &__caret {
    display: inline-flex;
    align-items: center;
    margin-left: auto;
    color: $color--primary;
    transition: transform $animate--time ease;
  }

  &__selected {
    display: flex;
    align-items: center;
    overflow: hidden;

    &-content {
      overflow: hidden;
      white-space: nowrap;
      text-overflow: ellipsis;
    }
  }

  &__items {
    display: flex;
    flex-direction: column;
    overflow: hidden;
    margin: 0;
    padding-top: $size--12;
    max-height: 200px;
    transition: all $animate--time;
  }

  &__item {
    flex-shrink: 0;
    overflow: hidden;
    padding: ($size--20 / 2) $size--16;
    word-wrap: break-word;
    text-overflow: ellipsis;
    transition: opacity $animate--time;
    cursor: pointer;

    &:hover {
      color: $color--primary-accent;
      background-color: $color--complimentary-bg-light;
    }

    &--selected {
      color: $color--primary;
    }

    &--multiple {
      overflow: hidden;
      word-wrap: break-word;
      text-overflow: ellipsis;
    }
  }

  &--rounded {
    padding-top: 0;

    .s-select {
      &__wrapper {
        border: 1px solid $color--border;
        border-radius: $size--24;
      }

      &__toggler {
        padding: 0 $size--16;
      }
    }
  }

  &--disabled {
    .s-select {
      &__wrapper {
        color: $color--font-secondary;
      }
    }

    &.s-select--rounded {
      .s-select__wrapper {
        border-color: $color--background;
        background-color: $color--background;
      }
    }
  }

  &--opened {
    position: relative;

    &:not(.s-select--rounded) {
      margin-right: -$size--16;
      margin-left: -$size--16;
      width: auto;

      .s-select__toggler {
        padding: 0 $size--16;
      }
    }

    .s-select {
      &__wrapper {
        @include z-index(dropdown);

        border-color: transparent;
        border-radius: $size--24 $size--24 0 0;
        box-shadow: 0 0 0 3px $color--outline;
      }

      &__search {
        max-height: $size--40;
        opacity: 1;
        visibility: visible;
      }

      &__empty {
        opacity: 1;
        visibility: visible;
      }

      &__caret {
        transform: rotate(180deg);
      }
    }

    .s-select__toggler {
      .svg-icon {
        transform: rotate(180deg);
      }
    }
  }

  &:not(&--opened):not(&--disabled) &__wrapper:hover {
    border-color: $color--primary-accent;
  }
}
