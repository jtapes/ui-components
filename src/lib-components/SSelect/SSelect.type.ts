export type SSelectItem<T> = {
  text: string
  value: T
}

export type SSelectItemValue = { [key: string]: any } | string | number
