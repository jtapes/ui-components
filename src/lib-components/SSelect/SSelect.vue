<template>
  <div
    class="s-select"
    :class="[
      `s-select--${state}`,
      {
        's-select--opened': opened,
        's-select--rounded': rounded,
      },
    ]"
  >
    <div ref="wrapper" class="s-select__wrapper">
      <div class="s-select__toggler" @click="handleOpenedChange">
        <s-input
          :value="currentValue"
          v-bind="$attrs"
          :state="state !== 'disabled' ? 'readonly': 'disabled'"
          :rounded="rounded"
          :placeholder="placeholder"
        >
          <template v-if="state !== 'disabled'" #append>
            <svg-icon name="dropdown-24-32" class="icon-32" />
          </template>
        </s-input>
      </div>
      <transition name="s-zoom-in-top">
        <div
          v-show="opened"
          ref="dropdown"
          class="s-select__dropdown text-sm"
          :class="{
            's-select__dropdown--opened': opened,
            's-select__dropdown--appended-to-body': appendToBody,
          }"
          :style="{ ...appendedToBodyPosition, ...appendedToBodySize }"
        >
          <div v-if="searchable" class="s-select__search">
            <s-input v-model="searchText" placeholder="Искать">
              <template #append>
                <svg-icon name="search-24-32" class="icon-32"/>
              </template>
            </s-input>
          </div>
          <div class="s-select__items">
            <template v-if="filteredItems.length">
              <template v-if="multiple">
                <s-checkbox
                  v-for="(item, index) in searchable ? filteredItems : mappedItems"
                  :key="index"
                  class="s-select__item ai-fs"
                  :value="value"
                  :checked-value="item.value"
                  multiple
                  @change="handleSelectedChange"
                >
                  <span class="s-select__item--multiple">{{ item.text }}</span>
                </s-checkbox>
              </template>
              <div
                v-for="(item, index) in searchable ? filteredItems : mappedItems"
                v-else
                :key="index"
                class="s-select__item"
                :class="{
                's-select__item--selected bold': item.value === value,
              }"
                @click="handleSelectedChange(item.value)"
              >
                {{ item.text }}
              </div>
            </template>
            <div v-else class="s-select__empty">
              <span class="color-font-muted">Ничего не найдено</span>
            </div>
          </div>
        </div>
      </transition>
    </div>
  </div>
</template>

<script lang="ts">
import { Component, Vue, Prop, Model } from 'vue-property-decorator'
import SInput from '@/lib-components/SInput/SInput.vue'
import SButton from '@/lib-components/SButton/SButton.vue'
import SCheckbox from '@/lib-components/SCheckbox/SCheckbox.vue'
import SvgIcon from '@/lib-components/SvgIcon/SvgIcon.vue'

import type { SSelectItem, SSelectItemValue } from '@/lib-components/SSelect/SSelect.type'

type AppendedToBodySize = {
  width?: string
}

type AppendedToBodyPosition = {
  top?: string
  left?: string
}

@Component({
  components: {
    SInput,
    SButton,
    SCheckbox,
    SvgIcon,
  },
})

export default class SSelect extends Vue {
  opened = false
  searchText = ''

  appendedToBodyPosition: AppendedToBodyPosition = {}
  appendedToBodySize: AppendedToBodySize = {}

  @Model('change')
  value!: SSelectItemValue

  /** Массив данных для контента */
  @Prop({ type: Array, default: () => [] })
  items!: SSelectItem<SSelectItemValue>[] | number[] | string[]

  /** Placeholder для селекта */
  @Prop({ type: String, default: 'Выберите' })
  placeholder!: string

  /** Название свойства для лэйбла */
  @Prop({ type: String, default: 'text' })
  itemText!: string

  /** Название свойства для значения */
  @Prop({ type: String, default: 'value' })
  itemValue!: string

  /** Отвечает за возможность выбирать множество значений */
  @Prop({ type: Boolean, default: false })
  multiple!: boolean

  /** Отвечает за возможность поиска в селекте */
  @Prop({ type: Boolean, default: false })
  searchable!: boolean

  /** Состояние радио ("disabled", "error") */
  @Prop({ type: String, default: 'default' })
  state!: string

  /** Свойство округления селекта */
  @Prop({ type: Boolean, default: true })
  rounded!: boolean

  /** Свойство выноса списка в body документа */
  @Prop({ type: Boolean, default: false })
  appendToBody!: boolean

  get mappedItems(): SSelectItem<SSelectItemValue>[] {
    return this.items.map((item: SSelectItemValue) => {
      const text = typeof item === 'object' ? item[this.itemText] : String(item)
      const value = typeof item === 'object' ? item[this.itemValue] : item
      return {
        text,
        value,
      }
    })
  }

  get filteredItems(): SSelectItem<SSelectItemValue>[] {
    return this.mappedItems.filter((item: SSelectItem<SSelectItemValue>) =>
      item.text.toLowerCase().includes(this.searchText.toLowerCase())
    )
  }

  get currentValue(): string {
    if (this.multiple && this.value && Array.isArray(this.value)) {
      return this.value.length
        ? `Выбрано ${this.value.length} ${this.declination}`
        : ''
    }

    const index = this.mappedItems.findIndex(
      (item) => item.value === this.value
    )
    return index !== -1 ? this.mappedItems[index].text : ''
  }

  get declination(): string {
    const cases = [2, 0, 1, 1, 1, 2]
    let number = this.multiple && this.value && Array.isArray(this.value) ? this.value.length : 0
    const words = ['значение', 'значения','значений']
    return words[
      number % 1 !== 0
        ? 1
        : number % 100 > 4 && number % 100 < 20
          ? 2
          : cases[number % 10 < 5 ? number % 10 : 5]
      ]
  }

  handleOpenedChange(): void {
    if (this.state !== 'disabled') {
      this.opened = !this.opened
      this.searchText = ''
      if (this.opened && this.appendToBody) {
        this.prepareDropdown()
      }
    }
  }

  prepareDropdown(): void {
    if (!this.$isServer) {
      this.$nextTick(async () => {
        if (this.$refs.wrapper && this.$refs.dropdown) {
          const slotRect = (this.$refs.wrapper as Element).getBoundingClientRect()
          this.appendedToBodySize = {
            width: `${slotRect.width}px`,
          }
          this.appendedToBodyPosition = {
            top: `${slotRect.top + slotRect.height + window.scrollY}px`,
            left: `${slotRect.left}px`,
          }
        }
      })
    }
  }

  close(e: MouseEvent): void {
    const element = e.target as Node

    if (!(this.$refs.dropdown as Node).contains(element) && !this.$el.contains(element)) {
      this.opened = false
      this.searchText = ''
    }
  }

  handleSelectedChange(value: SSelectItemValue): void {
    if (!this.multiple) {
      this.handleOpenedChange()
    }
    this.$emit('change', value)
  }

  mounted(): void {
    document.addEventListener('click', this.close, true)
    if (this.appendToBody) {
      document.body.append((this.$refs.dropdown as Node))
      this.prepareDropdown()
    }
  }

  beforeDestroy(): void {
    document.removeEventListener('click', this.close, true)
  }
}
</script>
