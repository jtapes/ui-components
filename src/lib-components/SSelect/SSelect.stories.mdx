import { action } from '@storybook/addon-actions'
import { config, withDesign } from 'storybook-addon-designs'
import { withTests } from '@storybook/addon-jest';
import {
  Meta,
  Story,
  ArgsTable,
  Description,
  Source,
  Props,
  Preview,
} from '@storybook/addon-docs'
import SSelect from '@/lib-components/SSelect/SSelect.vue'
import results from '/.jest-test-results.json';

<Meta
  title="DESIGN SYSTEM/Atom/Select"
  component={SSelect}
  decorators={[withDesign, withTests({ results })]}
  argTypes={{
    state: {
      control: { type: 'select' },
      options: [
        'disabled',
        'error',
        'default',
      ]
    },
    multiple: { control: 'boolean' },
    searchable: { control: 'boolean' },
    items: {
      control: 'array',
      defaultValue: [
        { text: 'Значение 1', value: 1 },
        { text: 'Значение 2', value: 2 },
        { text: 'Значение 3', value: 3 }
      ]
    },
    placeholder: { control: 'text' },
    itemText: { control: 'text' },
    itemValue: { control: 'text' },
    rounded: { control: 'boolean', defaultValue: true },
    appendToBody:  { control: 'boolean', defaultValue: true },
  }}
  parameters={{
    design: config({
      type: 'figma',
      url:
        'https://www.figma.com/file/IIeFkxNoHGcaqNHlVeNrhW/MVP-3-%2F-Market-%2F-UI-Kit?node-id=846%3A572',
    }),
    docs: {
      source: {
        type: 'code',
      },
    },
  }}
/>

# Выпадающий список

<Description of={SSelect} />

export const Template = (args, { argTypes }) => ({
  props: Object.keys(argTypes),
  components: { SSelect },
  template:
    `
      <div style="width: 300px; margin-left: 100px; height: 300px;">
        <SSelect
          :state="state"
          :multiple="multiple"
          :searchable="searchable"
          :items="items"
          :placeholder="placeholder"
          :item-text="itemText"
          :item-value="itemValue"
          :rounded="rounded"
          :appendToBody="appendToBody"
          v-model="selected"
          @change="handleChange"
        />
      </div>
`,
    data() {
        return {
            selected: null
        }
    },
    methods: {
        handleChange() {
            action('change')()
        }
    }
})

<Preview withToolbar withSource="none">
  <Story name="SSelect" args={{ searchable: true }}>
    {Template.bind({})}
  </Story>
</Preview>

## Пример

<Source
  language="html"
  code={`
    <s-select v-model="selectedItem" searchable/>
  `}
/>

<ArgsTable of={SSelect} />
