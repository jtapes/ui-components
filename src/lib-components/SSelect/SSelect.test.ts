import { mount } from '@vue/test-utils'
import SSelect from '@/lib-components/SSelect/SSelect.vue'

describe('SSelect', () => {
  const items = ['Значение 1', 'Значение 2']

  test('Is Vue snapshot', () => {
    const wrapper = mount(SSelect, {
      propsData: {
        multiple: true,
        searchable: true,
        items,
      },
    })
    expect(wrapper.element).toMatchSnapshot()
  })
  test('Is emitted by $emit()', async () => {
    const wrapper = mount({
      template: `
        <s-select
          v-model="selectedValue"
          :items="items"
          :multiple="isMultiple"
          @change="change"
        />`,
      components: { SSelect },
      data: () => ({
        isMultiple: true,
        items,
        selectedValue: [],
      }),
      methods: {
        change(value) {
          this.$emit('change', value)
        },
      },
    })

    let selectItem = wrapper.find('input[type="checkbox"]')
    await selectItem.setChecked()
    await selectItem.trigger('change')

    await wrapper.setData({
      isMultiple: false,
    })
    selectItem = wrapper.find('.s-select__item')
    await selectItem.trigger('click')

    expect(wrapper.emitted('change')).toBeTruthy()
  })
  test('Close select when click on body', async () => {
    const wrapper = mount(SSelect, {
      propsData: {
        items,
      },
    })

    await wrapper.find('.s-select__toggler').trigger('click')
    await document.body.click()

    expect(wrapper.vm.$data.opened).toBe(false)
    wrapper.destroy()
  })
  test('Test with empty props', () => {
    const wrapper = mount(SSelect)

    expect(wrapper.vm.$props.items.length).toBe(0)
  })
})
