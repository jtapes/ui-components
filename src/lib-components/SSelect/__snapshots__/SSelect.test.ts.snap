// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SSelect Is Vue snapshot 1`] = `
<div
  class="s-select s-select--default s-select--rounded"
>
  <div
    class="s-select__wrapper"
  >
    <div
      class="s-select__toggler"
    >
      <div
        class="s-input s-input--text s-input--readonly s-input--rounded"
      >
        <!---->
         
        <input
          class="s-input__inner s-input__inner--appended"
          placeholder="Выберите"
          readonly="readonly"
          type="text"
        />
         
        <!---->
         
        <!---->
         
        <div
          class="s-input__append"
        >
          <svg
            class="svg-icon icon-32 svg-icon--dropdown-24-32"
            xmlns="http://www.w3.org/2000/svg"
          >
            <use
              href="#dropdown-24-32"
              xlink:href="#dropdown-24-32"
            />
          </svg>
        </div>
      </div>
    </div>
     
    <transition-stub
      name="s-zoom-in-top"
    >
      <div
        class="s-select__dropdown text-sm"
        style="display: none;"
      >
        <div
          class="s-select__search"
        >
          <div
            class="s-input s-input--text s-input--rounded"
          >
            <!---->
             
            <input
              class="s-input__inner s-input__inner--appended"
              placeholder="Искать"
              type="text"
            />
             
            <!---->
             
            <!---->
             
            <div
              class="s-input__append"
            >
              <svg
                class="svg-icon icon-32 svg-icon--search-24-32"
                xmlns="http://www.w3.org/2000/svg"
              >
                <use
                  href="#search-24-32"
                  xlink:href="#search-24-32"
                />
              </svg>
            </div>
          </div>
        </div>
         
        <div
          class="s-select__items"
        >
          <label
            class="s-checkbox text-sm s-select__item ai-fs s-checkbox--default"
          >
            <div
              class="s-checkbox__input"
            >
              <div
                class="s-checkbox__inner"
              >
                <svg
                  class="svg-icon icon-16 svg-icon--checked-16"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <use
                    href="#checked-16"
                    xlink:href="#checked-16"
                  />
                </svg>
              </div>
               
              <input
                aria-hidden="false"
                class="s-checkbox__original"
                type="checkbox"
                value="Значение 1"
              />
            </div>
             
            <div
              class="s-checkbox__label"
            >
              <span
                class="s-select__item--multiple"
              >
                Значение 1
              </span>
            </div>
          </label>
          <label
            class="s-checkbox text-sm s-select__item ai-fs s-checkbox--default"
          >
            <div
              class="s-checkbox__input"
            >
              <div
                class="s-checkbox__inner"
              >
                <svg
                  class="svg-icon icon-16 svg-icon--checked-16"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <use
                    href="#checked-16"
                    xlink:href="#checked-16"
                  />
                </svg>
              </div>
               
              <input
                aria-hidden="false"
                class="s-checkbox__original"
                type="checkbox"
                value="Значение 2"
              />
            </div>
             
            <div
              class="s-checkbox__label"
            >
              <span
                class="s-select__item--multiple"
              >
                Значение 2
              </span>
            </div>
          </label>
        </div>
      </div>
    </transition-stub>
  </div>
</div>
`;
